django/contrib/postgres/constraints.py

```diff
@@ -0,0 +1,90 @@
+from django.db.models import CheckConstraint, F
+from django.db.models.sql import Query
+
+__all__ = ['ExcludeConstraint']
+
+
+class ExcludeConstraint(CheckConstraint):
+    template = 'CONSTRAINT %(name)s EXCLUDE USING %(index_type)s(%(expressions)s)%(where)s'
+
+    def __init__(self, *, expressions=(), index_type='GIST', operators=(), check=None, name=''):
+        if not len(expressions) == len(operators):
+            raise ValueError('Expressions and operators need to be of the same length.')
+
+        self.expressions = expressions
+        self.operators = operators
+        self.index_type = index_type
+        super().__init__(check=check, name=name)
+
+    def _parse_expressions(self, compiler, schema_editor, query):
+        expressions = []
+        for expression in self.expressions:
+            if isinstance(expression, (F, str)):
+                base = F(expression) if isinstance(expression, str) else expression
+                sql, params = base.resolve_expression(query=query, simple_col=True).as_sql(
+                    compiler=compiler, connection=schema_editor.connection
+                )
+            else:
+                sql, params = expression.as_sql(compiler, schema_editor.connection)
+            expressions.append(sql % params)
+
+        return [
+            expression + ' WITH ' + self.operators[inx]
+            for inx, expression in enumerate(expressions)
+        ]
+
+    def _parse_where(self, compiler, schema_editor, query):
+        connection = schema_editor.connection
+        if self.check is not None:
```

Comment:
I now realize that you used `check` for the _partial_ part of the constraint. I feel like we should use a different name here as this is likely what we'll want to add to `UniqueConstraint` as well and `check` is already taken by `CheckConstraint` and has a different meaning. I think we should favor `condition` as used in #10140 given it's actually passed to the underlying index as well.

https://www.postgresql.org/docs/11/static/sql-createtable.html#SQL-CREATETABLE-EXCLUDE

> The predicate allows you to specify an exclusion constraint on a subset of the table; internally this creates a partial index.